Userland/Libraries/LibGUI/TextDocument.cpp

```diff
@@ -384,6 +384,99 @@ DeprecatedString TextDocument::text_in_range(TextRange const& a_range) const
     return builder.to_deprecated_string();
 }
 
+// This function will return the position of the previous grapheme cluster
+// break, relative to the cursor, for "correct looking" parsing of unicode based
+// on grapheme cluster boundary algorithm.
+size_t TextDocument::get_previous_grapheme_cluster_boundary(
+    TextPosition const& cursor) const
+{
+    if (!cursor.is_valid())
+        return 0;
+
+    auto& line = this->line(cursor.line());
+    auto line_span = Span<u32 const>(line.code_points(), line.length());
+
+    if (line_span.is_empty())
+        return 0;
+
+    auto boundaries_cache = Unicode::find_grapheme_segmentation_boundaries(
+        Utf32View(line.code_points(), line.length()));
+    if (boundaries_cache.is_empty())
+        return cursor.column();
+
+    auto boundary_guess = boundaries_cache.at(0);
+
+    for (auto current_boundary_guess : boundaries_cache) {
+        // If the current boundary is further to the right, update our guess.
+        if (current_boundary_guess < cursor.column())
+            boundary_guess = current_boundary_guess;
+        else
+            break;
+    }
+
+    // Return the distance in code points from the cursor.
+    return boundary_guess;
+}
+
+// This function will return the position of the next grapheme cluster break,
+// relative to the cursor, for "correct looking" parsing of unicode based on
+// grapheme cluster boundary algorithm.
+size_t TextDocument::get_next_grapheme_cluster_boundary(
+    TextPosition const& cursor) const
+{
+    if (!cursor.is_valid())
+        return 0;
+
+    auto& line = this->line(cursor.line());
+    auto line_span = Span<u32 const>(line.code_points(), line.length());
+
+    if (line_span.is_empty())
+        return 0;
+
+    auto boundaries_cache = Unicode::find_grapheme_segmentation_boundaries(
+        Utf32View(line.code_points(), line.length()));
+
+    if (boundaries_cache.is_empty())
+        return cursor.column();
+
+    auto boundary_guess = boundaries_cache.at(1);
+
+    for (auto current_boundary_guess : boundaries_cache.in_reverse()) {
+        // If the current boundary is further to the right, update our guess.
+        if (current_boundary_guess < cursor.column())
+            boundary_guess = current_boundary_guess;
+        else
+            break;
+    }
+    // Assume that the cursor cannot sit at a point that is not a boundary, so
+    // return previous guess.
+    return boundary_guess;
+}
```

Comment:
The recent changes to master let us avoid the allocations / manual searching incurred by `find_grapheme_segmentation_boundaries` to get the one index we are interested in. Specifically, see `previous_grapheme_segmentation_boundary` and `next_grapheme_segmentation_boundary`.

These means these two functions can reduce down to:

```c++
size_t TextDocument::get_previous_grapheme_cluster_boundary(TextPosition const& cursor) const
{
    if (!cursor.is_valid())
        return 0;

    auto const& line = this->line(cursor.line());

    auto index = Unicode::previous_grapheme_segmentation_boundary(line.view(), cursor.column());
    return index.value_or(cursor.column() - 1);
}

size_t TextDocument::get_next_grapheme_cluster_boundary(TextPosition const& cursor) const
{
    if (!cursor.is_valid())
        return 0;

    auto const& line = this->line(cursor.line());

    auto index = Unicode::next_grapheme_segmentation_boundary(line.view(), cursor.column());
    return index.value_or(cursor.column() + 1);
}
```